src/OpenLoco/src/Map/MapGenerator.cpp

```diff
@@ -136,15 +142,261 @@ namespace OpenLoco::World::MapGenerator
+            }
+
+            // 0x00462718
+            const auto randX = (randomVal >> 14) & 0x1FF;
+            const auto randY = (randomVal >> 23) & 0x1FF;
+            if (randX < 2)
+            {
+                printf("Bail! x = %d < 2\n", randX);
+                return;
+            }
+            if (randY < 2)
+            {
+                printf("Bail! y = %d < 2\n", randY);
+                return;
+            }
+
+            // Vanilla would use an `rcl` for the lower bit, getting the carry flag in.
+            // We're substituting it with an extra bit of the random state.
+            const bool topographyBit = randX > randY;
+            const auto topographyId = (enumValue(options.topographyStyle) << 1) | (topographyBit ? 1U : 0);
+            const auto topographyFlags = kTopographyStyleFlags[topographyId];
+
+            const bool generateHills = (topographyFlags & TopographyFlags::hasHills) != TopographyFlags::none;
+            const bool generateMountains = (topographyFlags & TopographyFlags::hasMountains) != TopographyFlags::none;
+
+            const bool isValidHillIndex = generateHills && randomHillIndex < hillShapesObj->hillHeightMapCount;
+            const bool isValidMountainIndex = generateMountains && randomHillIndex >= hillShapesObj->hillHeightMapCount;
+            if (!(isValidHillIndex || isValidMountainIndex))
+            {
+                printf("Bail: flags = %d, randomHillIndex = %d, hillHeightMapCount = %d, mountainHeightMapCount = %d\n", enumValue(topographyFlags), randomHillIndex, hillShapesObj->hillHeightMapCount, hillShapesObj->mountainHeightMapCount);
+                return;
+            }
+
+            // 0x0046276D
+
+            if ((options.scenarioFlags & Scenario::ScenarioFlags::hillsEdgeOfMap) == Scenario::ScenarioFlags::none)
+            {
+                if (randX + featureWidth >= 382)
+                {
+                    printf("Edge of map, skipping\n");
+                    return;
+                }
+                if (randY + featureHeight >= 382)
+                {
+                    printf("Edge of map, skipping\n");
+                    return;
+                }
+            }
+
+            auto* src = g1Element->offset;
+
+            if (flipHillImageLateral)
+            {
+                if (flipHillImageVertical)
+                {
+                    int32_t x = randX;
+                    for (auto j = 0; j < featureWidth; ++j)
+                    {
+                        int32_t y = (featureHeight + randY - 1) & 0x1FF;
+                        for (auto i = 0; i < featureHeight; ++i)
+                        {
+                            const auto data = *src++;
+                            heightMap[Point{ x, y }] = std::max(data, heightMap[Point{ x, y }]);
+                            y--;
+                            y &= 0x1FF;
+                        }
+                        x++;
+                        x &= 0x1FF;
```

Comment:
vanilla didn't do this and i presume it would therefore have some odd hills if the map was allowed to go all the way to 511.